bson/flexible_bson/writer.py

```diff
@@ -0,0 +1,288 @@
+        self._update_state(new_state)
+
+    def end_document(self):
+        if self._state == _BSONWriterState.DOCUMENT:
+            new_state = None
+        elif self._state == _BSONWriterState.TOP_LEVEL:
+            new_state = _BSONWriterState.FINAL
+        else:
+            raise BSONWriterException(
+                "cannot end non-document or finished document.")
+
+        self._write_end_container()
+        self._update_state(new_state)
+
+    def start_array(self, name=None):
+        self._set_field_name(name)
+        self._write_start_container(BSONTypes.ARRAY)
+        self._update_state(_BSONWriterState.ARRAY)
+        self._array_indices.append(0)
+
+    def end_array(self):
+        if self._state == _BSONWriterState.ARRAY:
+            self._write_end_container()
+            self._update_state(None)
+            self._array_indices.pop()
+            return
+
+        raise BSONWriterException("cannot end non-array.")
+
+    def write_name(self, name):
+        self._set_field_name(name)
+
+    def write_value(self, value):
+        c_name = self._get_c_field_name()
+        self._insert_bytes(_name_value_to_bson(
+            c_name, value, self._check_keys, self._codec_options))
+
+    def write_name_value(self, name, value):
+        self.write_name(name)
+        self.write_value(value)
+
+
+class BSONFieldWriter(BSONWriter):
+    def __init__(self, field_name=None, *args, **kwargs):
+        # field_name is a c-string
+        if field_name is None:
+            raise BSONWriterException("must provide field name")
+        self._field_name = field_name
+        super(BSONFieldWriter, self).__init__(*args, **kwargs)
+
+    def _get_c_field_name(self):
+        if self._field_name is not None:
+            c_name = self._field_name
+            self._field_name = None
+            return c_name
+        return super(BSONFieldWriter, self)._get_c_field_name()
+
+    def as_bytes(self):
+        bstream = super(BSONFieldWriter, self).as_bytes()
+
+        # Drop 4 bytes for int32 size at the beginning.
+        # Drop 1 byte for '\x00' delimiter at the end.
+        return bstream[4:-1]
+
+
+class BSONDocumentWriter(BSONWriter):
+    def __init__(self, id_generator=ObjectId, *args, **kwargs):
+        if not id_generator or not callable(id_generator):
+            raise BSONWriterException("must supply id generator callable")
+        self._id_generator = id_generator
+        super(BSONDocumentWriter, self).__init__(*args, **kwargs)
+
+    def initialize(self):
+        # Flag that tracks whether we have written the `_id`.
+        self._id_written = False
+
+        # Field to store the `_id` value.
+        self._id = None
+
+        # Initialize basic BSONWriter.
+        super(BSONDocumentWriter, self).initialize()
+
+    def get_document_id(self):
+        return self._id
+
+    def set_document_id(self, id):
+        self._id = id
+
+    def _get_field_name(self):
```

Comment:
This method name is a bit confusing, since it doesn't just return the name, it also writes it into the BSON stream.